test/hotspot/jtreg/serviceability/sa/TestObjectMonitorIterate.java

```diff
@@ -0,0 +1,97 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import sun.jvm.hotspot.HotSpotAgent;
+import sun.jvm.hotspot.oops.Oop;
+import sun.jvm.hotspot.runtime.ObjectMonitor;
+import sun.jvm.hotspot.runtime.ObjectSynchronizer;
+import sun.jvm.hotspot.runtime.VM;
+
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.SA.SATestUtils;
+
+/**
+ * @test
+ * @bug 8259008
+ * @library /test/lib
+ * @requires vm.hasSA
+ * @modules jdk.hotspot.agent/sun.jvm.hotspot
+ *          jdk.hotspot.agent/sun.jvm.hotspot.oops
+ *          jdk.hotspot.agent/sun.jvm.hotspot.runtime
+ * @run main TestObjectMonitorIterate
+ */
+
+public class TestObjectMonitorIterate {
+
+    private static void test(String pid) {
+        HotSpotAgent agent = new HotSpotAgent();
+        agent.attach(Integer.parseInt(pid));
+        try {
+            var heap = VM.getVM().getObjectHeap();
+            var itr = ObjectSynchronizer.objectMonitorIterator();
+            while (itr.hasNext()) {
+                ObjectMonitor mon = (ObjectMonitor)itr.next();
+                Oop oop = heap.newOop(mon.object());
+                System.out.println("Monitor found: " + oop.getKlass().getName().asString());
+                return;
+            }
+        } finally {
+            agent.detach();
+        }
+        throw new RuntimeException("Monitor not found");
+    }
+
+    private static void createAnotherToAttach(long lingeredAppPid) throws Exception {
+        // Start a new process to attach to the lingered app
+        ProcessBuilder processBuilder = ProcessTools.createJavaProcessBuilder(
+            "--add-modules=jdk.hotspot.agent",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.oops=ALL-UNNAMED",
+            "--add-exports=jdk.hotspot.agent/sun.jvm.hotspot.runtime=ALL-UNNAMED",
+            "TestObjectMonitorIterate",
+             Long.toString(lingeredAppPid));
+        SATestUtils.addPrivilegesIfNeeded(processBuilder);
+        OutputAnalyzer SAOutput = ProcessTools.executeProcess(processBuilder);
+        SAOutput.shouldHaveExitValue(0);
```

Comment:
Is it reasonable to expect at least one ObjectMonitor to be found?
There are a couple of system threads that wait on ObjectMonitors
so if the target VM runs for a little bit of time, there will be at least
two contended ObjectMonitors.

Reply:
Yes, we can see the monitor for reference queue at least.
This test aims to check whether SA can find the monitors.

Reply:
I wasn't trying to state that you should only check for one ObjectMonitor.
I was trying to make sure that I understood your logic and that you are
intentionally requiring at least one ObjectMonitor (and you are).

For diagnostic purposes, it would be good if you dumped all the ObjectMonitors
that were found. That would require switching from the if-statement back to
the while-loop, setting a variable when you find an ObjectMonitor and then
doing the return outside the while-loop when you know that you've seen an
ObjectMonitor.